firewood/src/merkle/parallel.rs

```diff
@@ -0,0 +1,485 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE.md for licensing terms.
+
+use crate::db::BatchOp;
+use crate::merkle::{Key, Merkle, Value};
+use crate::v2::api::KeyValuePairIter;
+use firewood_storage::logger::error;
+use firewood_storage::{
+    BranchNode, Child, FileBacked, FileIoError, ImmutableProposal, LeafNode, MaybePersistedNode,
+    MutableProposal, NibblesIterator, Node, NodeReader, NodeStore, Parentable, Path,
+};
+use rayon::{ThreadPool, ThreadPoolBuilder};
+use std::array::from_fn;
+use std::iter::once;
+use std::ops::Deref;
+use std::sync::mpsc::{Receiver, SendError, Sender};
+use std::sync::{Arc, OnceLock, mpsc};
+
+#[derive(Debug)]
+struct WorkerSender(mpsc::Sender<Request>);
+
+impl std::ops::Deref for WorkerSender {
+    type Target = mpsc::Sender<Request>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// A request to the worker.
+#[derive(Debug)]
+enum Request {
```

Comment:
This should be a `BatchOp<Key, Value>`. Then you get the convenience method `key()` which you can use later.